networkit/cpp/generators/ClusteredRandomGraphGenerator.cpp

```diff
@@ -3,55 +3,128 @@
+
 #include <networkit/auxiliary/Random.hpp>
 #include <networkit/generators/ClusteredRandomGraphGenerator.hpp>
-#include <networkit/structures/Partition.hpp>
+#include <networkit/graph/GraphBuilder.hpp>
 
 namespace NetworKit {
 
-ClusteredRandomGraphGenerator::ClusteredRandomGraphGenerator(count n, count k, double pin, double pout) : n(n), k(k), pin(pin), pout(pout) {
+ClusteredRandomGraphGenerator::ClusteredRandomGraphGenerator(count n, count k, double pIntra,
+                                                             double pInter)
+    : n(n), k(k), pIntra(pIntra), pInter(pInter), zeta(n) {
+
+    if (k == 0)
+        throw std::runtime_error("Error: k must be positive");
+
+    const auto isProbability = [](double p) noexcept -> bool { return (p >= 0) && (p <= 1); };
+
+    if (!isProbability(pIntra))
+        throw std::runtime_error("Error: pIntra must be in [0, 1]");
+
+    if (!isProbability(pInter))
+        throw std::runtime_error("Error: pInter must be in [0, 1]");
+
+    if (pIntra < pInter)
+        WARN("Intra-cluster probability is lower than inter-cluster probability.");
+
+    zeta.setUpperBound(k);
 }
 
 Graph ClusteredRandomGraphGenerator::generate() {
-    assert(pin >= pout);
+    GraphBuilder graphBuilder(n);
 
-    Graph G(n);
-    // assign nodes evenly to clusters
-    Partition zeta(n);
-    zeta.setUpperBound(k);
-    G.forNodes([&](node v){
-        index c = Aux::Random::integer(k-1);
-        zeta.addToSubset(c, v);
-    });
-
-    assert (zeta.numberOfSubsets() == k);
-
-    G.forNodePairs([&](node u, node v){
-        if (zeta.subsetOf(u) == zeta.subsetOf(v)) {
-            if (Aux::Random::probability() <= pin) {
-                G.addEdge(u, v);
-            }
-        } else {
-            if (Aux::Random::probability() <= pout) {
-                G.addEdge(u, v);
-            }
+    // Sequence of all nodes grouped by cluster. Here the vertices of each cluster are stored
+    // sequentially one after the other
+    std::vector<node> clustersSequence;
+
+    // Index in `clustersSequence` where each cluster ends. E.g., clusterEnd[i] is the index in
+    // `clustersSequence` of the first vertex in the (i + 1)-th cluster.
+    std::vector<size_t> clusterEnd(k);
+
+    {
+        // To assign vertices to clusters uniformly at random
+        std::uniform_int_distribution<index> clusterDistr{0, k - 1};
+
+        // Nodes grouped by cluster
+        std::vector<std::vector<node>> clusters(k);
+
+        std::vector<omp_lock_t> locks(k);
+        for (index i = 0; i < k; ++i)
+            omp_init_lock(&locks[i]);
+
+        // Assign nodes to clusters
+        graphBuilder.parallelForNodes([&zeta = zeta, &clusters, &clusterDistr, &locks](node u) {
+            const index cluster = clusterDistr(Aux::Random::getURNG());
+            zeta.addToSubset(cluster, u);
+            omp_set_lock(&locks[cluster]);
+            clusters[cluster].push_back(u);
+            omp_unset_lock(&locks[cluster]);
+        });
```

Comment:
This is basically a parallel counting sort. Two thoughts:

* What about applying a known and tested parallel counting sort algorithm?
* This most likely doesn't dominate the running time. The actual assignment (using `addToSubset` could still be parallelized as this might actually need some time to generate the random numbers. However, for simplicity I would suggest to instead implement the actual sorting sequentially without the additional `clusters` vector unless it turns out that this is actually a bottleneck.

Note that this implementation creates quite some overhead:
* We initialize and use locks, even if just a single thread is used
* We basically sort the nodes twice, once in the clusters vector and then in the final `clusterSequence` where actually just the sizes of the clusters would be needed to sort the nodes into `clusterSequence`.
* The `clusters` vector actually causes `1 + k * log(n/k)` memory allocations which could be completely avoided.

I wouldn't be surprised if sequentially this was a factor 2 to 4 slower than a simple sequential counting sort implementation and needed at least 8 cores to be as fast as a simple sequential counting sort implementation, if it becomes faster at all as randomly chosen cluster ids most likely cause heavy lock contention.